src/psyclone/psyir/backend/language_writer.py

```diff
@@ -0,0 +1,252 @@
+        :rtype: list of str
+
+        '''
+        raise NotImplementedError("gen_dims() is abstract")
+
+    # ------------------------------------------------------------------------
+    def arrayreference_node(self, node):
+        '''This method is called when an ArrayReference instance is found
+        in the PSyIR tree.
+
+        :param node: an ArrayNode PSyIR node.
+        :type node: :py:class:`psyclone.psyir.nodes.ArrayNode`
+
+        :returns: the code as a string.
+        :rtype: str
+
+        :raises VisitorError: if the node does not have any children.
+
+        '''
+        if not node.children:
+            raise VisitorError(
+                "Incomplete ArrayReference node (for symbol '{0}') found: "
+                "must have one or more children.".format(node.name))
+        args = self.gen_dims(node.children, node.name)
+        result = "{0}{2}{1}{3}".format(node.name,
+                                       ",".join(args),
+                                       self._array_parenthesis[0],
+                                       self._array_parenthesis[1])
+        return result
+
+    # ------------------------------------------------------------------------
+    def structurereference_node(self, node):
+        '''
+        Creates the code for an access to a member of a structure type.
+
+        :param node: a StructureReference PSyIR node.
+        :type node: :py:class:`psyclone.psyir.nodes.StructureReference`
+
+        :returns: the code as string.
+        :rtype: str
+
+        :raises VisitorError: if this node does not have an instance of Member\
+                              as its only child.
+
+        '''
+        if len(node.children) != 1:
+            raise VisitorError(
+                "A StructureReference must have a single child but the "
+                "reference to symbol '{0}' has {1}.".format(
+                    node.name, len(node.children)))
+        if not isinstance(node.children[0], Member):
+            raise VisitorError(
+                "A StructureReference must have a single child which is a "
+                "sub-class of Member but the reference to symbol '{0}' has a "
+                "child of type '{1}'".format(node.name,
+                                             type(node.children[0]).__name__))
+        result = node.symbol.name + self._structure_character + \
+            self._visit(node.children[0])
+        return result
+
+    # ------------------------------------------------------------------------
+    def member_node(self, node):
+        '''
+        Creates the code for an access to a member of a derived type.
+
+        :param node: a Member PSyIR node.
+        :type node: :py:class:`psyclone.psyir.nodes.Member`
+
+        :returns: the code as string
+        :rtype: str
+
+        '''
+        result = node.name
+        if not node.children:
+            return result
+
+        if isinstance(node.children[0], Member):
```

Comment:
Can you add some comments in each branch of this block to provide some context of the PSyIR structure each branch is dealing with?

Reply:
Done.